plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java

```diff
@@ -16,27 +16,82 @@
 
     @Override
-    public TestFramework createRetrying(TestFrameworkTemplate template, TestNames failedTests) {
-        TestFilterBuilder filters = template.filterBuilder();
-        addFilters(filters, template.testsReader, failedTests, false);
-        JUnitPlatformTestFramework newFramework = new JUnitPlatformTestFramework(filters.build());
-        copyTestOptions((JUnitPlatformOptions) template.task.getTestFramework().getOptions(), newFramework.getOptions());
-        return newFramework;
+    public TestFramework createRetrying(TestFrameworkTemplate template, TestFramework testFramework, TestNames failedTests) {
+        DefaultTestFilter failedTestsFilter = testFilterFor(failedTests, false, template);
+        return testFrameworkProvider(template, testFramework).testFrameworkFor(failedTestsFilter);
     }
 
-    private void copyTestOptions(JUnitPlatformOptions source, JUnitPlatformOptions target) {
-        target.setIncludeEngines(source.getIncludeEngines());
-        target.setExcludeEngines(source.getExcludeEngines());
-        target.setIncludeTags(source.getIncludeTags());
-        target.setExcludeTags(source.getExcludeTags());
+    interface TestFrameworkProvider {
+        class ProviderForCurrentGradleVersion implements TestFrameworkProvider {
+
+            private final TestFramework testFramework;
+
+            ProviderForCurrentGradleVersion(TestFramework testFramework) {
+                this.testFramework = testFramework;
+            }
+
+            @Override
+            public TestFramework testFrameworkFor(DefaultTestFilter failedTestsFilter) {
+                return testFramework.copyWithFilters(failedTestsFilter);            }
+        }
+
+        class ProviderForGradleOlderThanV8 implements TestFrameworkProvider {
+
+            private final TestFrameworkTemplate template;
+
+            ProviderForGradleOlderThanV8(TestFrameworkTemplate template) {
+                this.template = template;
+            }
+
+            @Override
+            public TestFramework testFrameworkFor(DefaultTestFilter failedTestsFilter) {
+                JUnitPlatformTestFramework retryTestFramework = newInstance(failedTestsFilter);
+                copyOptions((JUnitPlatformOptions) template.task.getTestFramework().getOptions(), retryTestFramework.getOptions());
+
+                return retryTestFramework;
+            }
+
+            private static JUnitPlatformTestFramework newInstance(DefaultTestFilter failedTestsFilter) {
+                try {
+                    Class<?> jUnitPlatformTestFrameworkClass = JUnitPlatformTestFramework.class;
+                    Constructor<?> constructor = jUnitPlatformTestFrameworkClass.getConstructor(DefaultTestFilter.class);
+
+                    return (JUnitPlatformTestFramework) constructor.newInstance(failedTestsFilter);
+                } catch (ReflectiveOperationException e) {
+                    throw new RuntimeException(e);
```

Comment:
Don't we have `UncheckedException` for this, or some other more fitting type?

Reply:
I've used the same exception as we're currently throwing in similar situations like the [RetryTestResultProcessor](https://github.com/gradle/test-retry-gradle-plugin/blob/main/plugin/src/main/java/org/gradle/testretry/internal/executer/RetryTestResultProcessor.java#L139-L141).

I think it makes sense to always throw the same here and therefore I suggest to keep it that way in order to minimize risk that we're breaking something else with customers. Personally, I probably would have chosen `UnsupportedOperationException`.

@leonard84 WDYT?